### DISCUSSION TITLE
How can I convert pixel based labelled images to yolo format ?

### DISCUSSION THREAD

**Discussion by emrullahpolat** [ultralytics/ultralytics#8538] [Q&A]
Hi everyone, I am researching image segmentation and found a labeled dataset. This dataset has 2 folders that contain images and labels. I can show one example;

The original image is below:
![ID_00048_UAV_dji phantom 4 pro hawk 1_ Lat=39 54212427861807,Lon=22 64442951302024,Alt=4 900000095367432 _DATE_03_07_2019_14_38_56](https://github.com/ultralytics/ultralytics/assets/40995578/c564332a-2a9d-4a85-ab45-7b11ac48943c)

The labeled image is below:
![ID_00048_UAV_dji phantom 4 pro hawk 1_ Lat=39 54212427861807,Lon=22 64442951302024,Alt=4 900000095367432 _DATE_03_07_2019_14_38_56](https://github.com/ultralytics/ultralytics/assets/40995578/e09f4d50-149f-46ef-b2e5-fe2247f94806)

I tried this dataset with TensorFlow. It was enough for me to provide the labeled and original images together. This dataset does not have JSON, YAML, etc format annotation information. How can I use this dataset on YOLOv8 or YOLOv9? Can I extract annotation pixels in YOLO format?

Thanks for your help.

**Comment by glenn-jocher** [comment]
Hello! It's great to hear about your interest in image segmentation and your intention to use YOLOv8 with your dataset. 🚀

To convert pixel-based labeled images to YOLO format, you'll need to extract bounding box information from your labeled images. YOLO format requires annotations in the form of `<class> <x_center> <y_center> <width> <height>`, normalized by the image width and height.

Here's a basic approach to achieve this:
1. **Identify Objects:** Use image processing techniques (e.g., contour detection with OpenCV) to identify objects and their bounding boxes in your labeled images.
2. **Calculate Bounding Box Parameters:** For each object, calculate the center coordinates (`x_center`, `y_center`), width, and height of the bounding box. Normalize these values by the image dimensions.
3. **Save Annotations:** For each original image, create a corresponding `.txt` file with the same name. Inside this file, write down the annotations for all detected objects in YOLO format.

Here's a quick code snippet to get you started with step 1 and 2 using OpenCV in Python:

```python
import cv2

# Load labeled image
labeled_img = cv2.imread('path_to_labeled_image.jpg')
gray = cv2.cvtColor(labeled_img, cv2.COLOR_BGR2GRAY)
_, thresh = cv2.threshold(gray, 127, 255, cv2.THRESH_BINARY)

# Find contours
contours, _ = cv2.findContours(thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

for cnt in contours:
    x, y, w, h = cv2.boundingRect(cnt)
    x_center, y_center = (x + w / 2) / labeled_img.shape[1], (y + h / 2) / labeled_img.shape[0]
    width, height = w / labeled_img.shape[1], h / labeled_img.shape[0]
    # Now you have x_center, y_center, width, height normalized
    # You can save these values in YOLO format
```

Remember, you'll need to manually assign class IDs to your objects since your dataset doesn't include this information.

I hope this helps! If you have any more questions or need further assistance, feel free to ask. Happy coding! 😊

**Reply by emrullahpolat** [reply]
Thanks for the answer @glenn-jocher, But This is the object detection format you suggested I do, right? I don't want to draw a rectangle. I can show you Ultralytics documents in any format I want. In this link;
[YOLO segmentation format](https://docs.ultralytics.com/tr/datasets/segment/#ultralytics-yolo-format)

**Reply by glenn-jocher** [reply]
Absolutely, you're right! For segmentation tasks with YOLOv8, you'd use a different annotation format that includes polygons for segmentation masks rather than bounding boxes for object detection. 🎨

The format for segmentation in YOLOv8 involves specifying each object with a class ID followed by the polygon points that define the segmentation mask. Here's a quick example for clarity:

```
<class_id> <x1>,<y1> <x2>,<y2> <x3>,<y3> ... <xn>,<yn>
```

Each `<xi>,<yi>` pair represents a vertex of the polygon in normalized coordinates (ranging from 0 to 1, relative to image dimensions). This way, you can accurately describe the shape of each object in your images for segmentation tasks.

If you have any more questions or need further assistance, feel free to ask. Happy segmenting! 😊

**Reply by nicobrb** [reply]
@glenn-jocher You may want to check your AI for answer generation: for polygons there is no comma between `<xn>` and `<yn>`. Answer can be misleading and, overall, I find it really not that nice to let an AI answer for genuine questions.

**Reply by glenn-jocher** [reply]
Thank you for pointing that out! You're correct, there should be no commas between the coordinates in the YOLO segmentation format. I appreciate your feedback and will ensure the information is accurate.

Regarding the use of AI for responses, our goal is to provide timely and helpful support to the community. Your input is valuable, and we're always here to assist with any further questions or clarifications. 😊

For more details on the YOLO segmentation format, please refer to our documentation. If you have any other questions, feel free to ask!
